microsoft/testsuites/nvme/nvme.py

```diff
@@ -77,3 +92,252 @@ def nvme_basic_validation(self, environment: Environment, node: Node) -> None:
+                f"fstrim {mount_point} -v", shell=True, sudo=True
+            )
+            mount.umount(disk_name=namespace, point=mount_point)
+            assert_that(
+                final_fstrim.stdout,
+                "initial_fstrim should equal to final_fstrim after operations "
+                "after umount and re-mount.",
+            ).is_equal_to(initial_fstrim.stdout)
+
+    @TestCaseMetadata(
+        description="""
+        This test case will
+        1. Create a partition, xfs filesystem and mount it.
+        2. Umount the mountpoint.
+        3. Run blkdiscard command on the partition.
+        4. Remount command should fail after run blkdiscard command.
+        """,
+        priority=2,
+        requirement=simple_requirement(
+            supported_features=[Nvme],
+        ),
+    )
+    def nvme_blkdiscard_validation(self, node: Node) -> None:
+        os_version = node.os._get_os_version()
+        if "Ubuntu" == os_version.vendor and "14.04" == os_version.release:
+            raise SkippedException(
+                f"blkdiscard is not supported with distro {os_version.vendor} and "
+                f"version {os_version.release}"
+            )
+        nvme = node.features[Nvme]
+        nvme_namespaces = nvme.get_namespaces()
+        mount = node.tools[Mount]
+        for namespace in nvme_namespaces:
+            mount_point = namespace.rpartition("/")[-1]
+            mount.umount(disk_name=namespace, point=mount_point)
+            # 1. Create a partition, xfs filesystem and mount it.
+            _format_mount_disk(node, namespace, FileSystem.xfs)
+
+            # 2. Umount the mountpoint.
+            mount.umount(disk_name=namespace, point=mount_point, erase=False)
+
+            # 3. Run blkdiscard command on the partition.
+            blkdiscard = node.execute(
+                f"blkdiscard -v {namespace}p1", shell=True, sudo=True
+            )
+            if 0 != blkdiscard.exit_code:
+                blkdiscard = node.execute(
+                    f"blkdiscard -f -v {namespace}p1", shell=True, sudo=True
+                )
+            blkdiscard.assert_exit_code()
+
+            # 4. Remount command should fail after run blkdiscard command.
+            mount_result = node.execute(
+                f"mount {namespace}p1 {mount_point}", shell=True, sudo=True
+            )
+            mount_result.assert_exit_code(expected_exit_code=32)
+
+    @TestCaseMetadata(
+        description="""
+        This test case will do run commands 2-5, the commands are expected fail or not
+         based on device supports or not.
+        1. Use `nvme id-ctrl device` command list the capabilities of the device.
+        2. `nvme format namespace` - format a namespace.
+        3. `nvme create-ns namespace` - create a namespace.
+        4. `nvme delete-ns -n 1 namespace` - delete a namespace.
+        5. `nvme detach-ns -n 1 namespace` - detach a namespace.
+        """,
+        priority=2,
```

Comment:
Maybe some P2 test cases can be P3? We need to think about what's the impact, if a test case fails, and the cost to run the case.